[win32] fix endless loop with fixed text metrics

akoch-yatta opened the pull request:
This commit fixes a regression that ignored bigger fixed font metrics in TextLayout in one scenario that led to an endloop recalculation loop.

Fixes #1610

fedejeanne commented on bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/TextLayout.java:
```diff
@@ -2137,8 +2137,9 @@ public FontMetrics getLineMetrics (int lineIndex) {
-	int ascentInPoints = this.ascent;
-	int descentInPoints = this.descent;
+	final int zoom = getZoom();
+	int ascentInPoints =  Math.max(DPIUtil.scaleDown(getDevice(), lptm.tmAscent, zoom), this.ascent);
```
Why not use the field `device` instead of calling `getDevice()` so often? I mean the call to `internal_new_GC` already checked if the device is disposed so starting in line 2132 you can safely assume the device is not disposed.

akoch-yatta replied:
Good point, Adapted all places in the changed method to use the field directly.

fedejeanne commented:
I can't understand the errors. I created an issue for it: #1683 

The rest seems to be fine but I would like to merge only after all tests ran, hopefully today. Fingers crossed 🤞

fedejeanne commented:
Ah, so close.... 

```java
java.lang.Error: SWT Resource was not properly disposed
	at org.eclipse.swt.graphics.Resource.initNonDisposeTracking(Resource.java:184)
	at org.eclipse.swt.graphics.Resource.<init>(Resource.java:114)
	at org.eclipse.swt.graphics.GC.<init>(GC.java:260)
	at org.eclipse.swt.graphics.GC.<init>(GC.java:228)
	at org.eclipse.swt.tests.junit.Test_org_eclipse_swt_custom_StyledText.test_bug1610_fixedLineHeightWithChangingToSmallerFont_noException(Test_org_eclipse_swt_custom_StyledText.java:5945)
...
```

@akoch-yatta there seems to be a leak in the test so I removed it from this PR and added another one for the test. We can merge that one later --> https://github.com/eclipse-platform/eclipse.platform.swt/pull/1687

fedejeanne commented:
Ignoring the usual suspects: https://github.com/eclipse-platform/eclipse.platform.swt/issues/1676 and https://github.com/eclipse-platform/eclipse.platform.swt/issues/1564

akoch-yatta commented:
> Ah, so close....
> 
> ```java
> java.lang.Error: SWT Resource was not properly disposed
> 	at org.eclipse.swt.graphics.Resource.initNonDisposeTracking(Resource.java:184)
> 	at org.eclipse.swt.graphics.Resource.<init>(Resource.java:114)
> 	at org.eclipse.swt.graphics.GC.<init>(GC.java:260)
> 	at org.eclipse.swt.graphics.GC.<init>(GC.java:228)
> 	at org.eclipse.swt.tests.junit.Test_org_eclipse_swt_custom_StyledText.test_bug1610_fixedLineHeightWithChangingToSmallerFont_noException(Test_org_eclipse_swt_custom_StyledText.java:5945)
> ...
> ```
> 
> @akoch-yatta there seems to be a leak in the test so I removed it from this PR and added another one for the test. We can merge that one later --> #1687

Damn, yes, forget to clean up at the end of the test. I will create a new PR for that test tomorrow
